content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -391,24 +400,209 @@ Here's the generic procedure again, now showing the overlay:
+padding bytes are considered parts of the components.
+
+To make that work, the compiler takes the most stringent alignment of
+all the record type's components and uses that for the alignment of the
+overall record type. That way, any address that satisfies the record
+object's alignment will satisfy the components' alignment requirements.
+The alignment is component X, of subtype S, is 4. The other components
+have an alignment of 1, therefore R'Alignment is 4. An aligned address
+plus 12 will also be an aligned address.
+
+This rounding up based on alignment is recommended behavior for the
+compiler, not a requirement, but is reasonable and typical among
+vendors. Although it can result in unused storage, that's the price paid
+for speed of access (or even correctness for machines that would fault
+on misaligned address allocations).
+
+
+As you can see, alignment is a critical factor in the sizes of composite
+objects. If you care about the layout of the type you very likely need
+to care about the alignment of the components and overall record type.
+
+Ada compilers are allowed to reorder the components of record types in
+order to minimize these gaps or satisfy the alignment requirements of
+the components. Some compilers do, some don't. Consider the type R
+again, this time with the first two components switched in the component
+declaration order:
+
+.. image:: images/optimized-record-component-order.png
+  :width: 600
+  :alt: Memory allocated to a record with optimized layout
+
+Now R'Size will report 56 bits instead of 80. The one trailing byte will
+still be padding, but only that one.
 
-There are other characteristics we might want to query too. For example, 
-we might want to ask the compiler what size and alignment it chose for a 
-given object (or type, for all such objects). Rather than describe all 
-the possibilities, we can just say that all the representation 
-characteristics that can be specified can also be queried. We cover 
-specifying representation characteristics next, so just assume the 
-corresponding query is available. 
+What about unbounded types, for example type String? Querying the 'Size
+in that case would provide an implementation-defined result. A somewhat
+silly thing to do, really, since the type |mdash| by definition |mdash|
+doesn't specify how many components are involved.
+
+Usually, though, you don't want to query the size of a type. Most of the
+time what you want is the size of objects of the type. Going back to
+sending values over the wire, the code should query the size of the
+*parameter* holding the value to be sent. That will tell you how many
+bits are really needed.
+
+One last point. GNAT, and now Ada 202x, define an attribute named
+Object_Size. It does just what the name suggests: what 'Size does when
+applied to objects rather than types. GNAT also defines another
+attribute, named Value_Size, that does what 'Size does when applied to
```

Comment:
```suggestion
:ada:`Object_Size`. It does just what the name suggests: what :ada:`'Size` does
when applied to objects rather than types. GNAT also defines another attribute, 
named :ada:`Value_Size`, that does what :ada:`'Size` does when applied to
```